src/Init/Data/Nat/Bitwise/Basic.lean

```diff
@@ -78,6 +78,8 @@ of a number.
 -/
 
 /-- `testBit m n` returns whether the `(n+1)` least significant bit is `1` or `0`-/
-def testBit (m n : Nat) : Bool := (m >>> n) &&& 1 != 0
+def testBit (m n : Nat) : Bool :=
+  -- `1 &&& n` is faster than `n &&& 1` for big `n`. This may change in the future.
+  1 &&& (m >>> n) != 0
```

Comment:
Do you have numbers to confirm this claim?
Is it faster in kernel reduction or for compiled code? If it really is faster, doesn't the compiler just optimize it?

Reply:
```lean
def test (x : Nat) : Nat := 1 &&& x
def test' (x : Nat) : Nat := x &&& 1

def run_test (n : Nat) (x : Nat) : IO Unit := do
  let mut t := 0
  for _ in [0 : n] do
    t := test x

def run_test' (n : Nat) (x : Nat) : IO Unit := do
  let mut t := 0
  for _ in [0 : n] do
    t := test' x

def n := 10^5
def x := 11^1000000

#eval timeit "1 &&& x" (run_test n x) -- 41ms
#eval timeit "x &&& 1" (run_test' n x) -- 1.5s
```

Reply:
That's a surprisingly big difference, almost suspiciously so. I only know enough about micro-benchmarking to know results can sometimes be misleading, so I tried playing with various variations (e.g., sprinkling `no_inline`) but everything points towards `1 &&& x` indeed being significantly faster.

Admittedly, In my original comment I was only thinking about "small" (up-to 64 bit) numbers, is this `x &&& 1` vs `1 &&& x` a known behaviour of gmp (which I assume what is going on here, with the big numbers given in your testcase)?